compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDataClassPrimaryConstructorChecker.kt

```diff
@@ -7,23 +7,38 @@ package org.jetbrains.kotlin.fir.analysis.checkers.declaration
 
 import org.jetbrains.kotlin.descriptors.ClassKind
 import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.hasValOrVar
 import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
 import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
 import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
-import org.jetbrains.kotlin.fir.declarations.FirConstructor
 import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.getPrimaryConstructorIfAny
 import org.jetbrains.kotlin.fir.declarations.isData
 
-object FirPrimaryConstructorRequiredForDataClassChecker : FirRegularClassChecker() {
+object FirDataClassPrimaryConstructorChecker : FirRegularClassChecker() {
     override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
         if (declaration.classKind != ClassKind.CLASS || !declaration.isData) {
             return
         }
 
-        val hasPrimaryConstructor = declaration.declarations.any { it is FirConstructor && it.isPrimary }
+        val primaryConstructor = declaration.getPrimaryConstructorIfAny()
```

Comment:
Well, I'm afraid the resolved comments didn't get enough attention: https://github.com/JetBrains/kotlin/pull/4143#discussion_r587675125